tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py

```diff
@@ -115,6 +118,257 @@ def test_serialize(self):
+            expected_input_device)
+
+    def check_device_argument_fallback(
+            self,
+            initial_model_device,
+            initial_input_device,
+            device_arg,
+            input_device_arg):
+
+        if self.omit_device:
+            device_arg = None
+        if self.omit_input_device:
+            input_device_arg = None
+
+        actual_converter_device_args = []
+
+        @chainer.dataset.converter()
+        def convert(arr, device):
+            # The converter records the given device.
+            actual_converter_device_args.append(device)
+            if device is None:
+                return arr
+            return device.send(arr)
+
+        class Model(chainer.Link):
+            def __init__(self):
+                chainer.Link.__init__(self)
+                with self.init_scope():
+                    self.p1 = chainer.Parameter()
+                    self.p2 = chainer.Parameter(
+                        numpy.array([1, 2], numpy.float32))
+
+            def forward(self, x):
+                return chainer.functions.identity(x)
+
+        model = Model()
+        model.to_device(initial_model_device)
+        optimizer = DummyOptimizer()
+        optimizer.setup(model)
+        iterator = DummyIterator([
+            initial_input_device.send(numpy.array(1)),
+        ])
+
+        # Make kwargs
+        kwargs = {}
+        if device_arg is not None:
+            kwargs['device'] = device_arg
+        if input_device_arg is not None:
+            kwargs['input_device'] = input_device_arg
+
+        # Calculate the expected devices
+        expect = self._get_expected_devices(
+            initial_model_device,
+            initial_input_device,
+            device_arg,
+            input_device_arg)
+
+        if expect is None:
+            # Error is expected
+            with pytest.raises(KeyError):
+                training.updaters.StandardUpdater(
+                    iterator, optimizer, convert, **kwargs)
+            return
+
+        (expected_device_attr,
+         expected_model_device,
+         expected_input_device) = expect
+
+        # Create the StandardUpdater
+        updater = training.updaters.StandardUpdater(
+            iterator, optimizer, convert, **kwargs)
+
+        assert updater.device == expected_device_attr
+        assert updater.input_device == expected_input_device
+
+        # Check the model device
+        if expected_model_device is None:
+            # Model device is unchanged
+            # TODO(niboshi): model.device should be initial_model_device too.
+            assert model.p1.device == initial_model_device
+            assert model.p2.device == initial_model_device
+        else:
+            # Model is transferred
+            # TODO(niboshi): model.device should be expected_model_device too.
+            assert model.p1.device == expected_model_device
+            assert model.p2.device == expected_model_device
```

Comment:
I minor but you can DRY this by conditionally assigning the RHS to a temporary.

Reply:
Fixed.